File: Baekjun/24.10.1/24232. 망가진 나무/solution.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().split("\n");
let line = 0

const N = Number(input[line++])

const G = Array.from(Array(N+1), ()=>[])

for (let i = 0; i < N-1; i++) {
    const [a, b] = input[line++].split(" ").map(Number)
    G[a].push([b,1,i])
    G[b].push([a,0,i])
}

const MAX = Number.MAX_VALUE

const dfs = () => {
    const V = Array(N+1).fill(MAX)
    const S = [1]
    let goal = 1
    V[1] = 0

    while (S.length > 0) {
        const n = S.pop()

        for (const [node, f, _] of G[n]) {
            if (V[node] < MAX) continue
            if (f === 0) {
                V[node] = V[n] - 1
            } else {
                V[node] = V[n] + 1
            }
            if(V[goal] > V[node]) {
                goal = node
            }
            S.push(node)
        }
    }
    return goal
}

const dfs2 = (goal) => {
    const edges = Array(N-1).fill('0')

    const S = [goal]
    const V = Array(N+1).fill(0)
    V[goal] = 1

    while (S.length > 0) {
        const n = S.pop()

        for (const [node, f, i] of G[n]) {
            if (V[node] === 1) continue
            V[node] = 1
            if (f === 0) {
                edges[i] = '1'
            }
            S.push(node)
        }
    }

    return edges
}

console.log(dfs2(dfs()).join(''));

File: Baekjun/24.10.1/24232. 망가진 나무/solution.py
import sys

input = sys.stdin.readline

N = int(input())
graph = [[] for _ in range(N + 1)]
order = dict()
edges = ["0"] * (N - 1)
for i in range(N - 1):
    u, v = map(int, input().split())
    graph[u].append((v, 1, i))
    graph[v].append((u, 0, i))


def dfs():
    S = [1]
    MAX = 10**6
    visited = [MAX] * (N + 1)
    visited[1] = 0

    goal = 1

    while S:
        now = S.pop()

        for node, f, i in graph[now]:
            if visited[node] != MAX:
                continue

            if f == 0:
                visited[node] = visited[now] - 1
            else:
                visited[node] = visited[now] + 1

            if visited[node] < visited[goal]:
                goal = node
            S.append(node)
    return goal


goal = dfs()


def dfs2():
    V = [0] * (N + 1)
    V[goal] = 1
    S = [goal]

    while S:
        n = S.pop()
        for node, f, i in graph[n]:
            if V[node]:
                continue

            V[node] = 1
            if f == 0:
                edges[i] = "1"

            S.append(node)


dfs2()

print("".join(edges))
